File: Hard/text_justification/text_justification.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
using namespace std;

class Solution {
    public:
        vector<string> fullJustify(vector<string>& words, int maxWidth) {
            vector<string> result;
            int i = 0;
            int n = words.size();

            while (i < n) {
                vector<string> line;
                int length = 0;

                // collect words for the line
                while (i < n && length + words[i].size() + line.size() <= maxWidth) {
                    line.push_back(words[i]);
                    length += words[i].size();
                    i++;
                }

                string s;

                // last line OR single word → left justify
                if (i == n || line.size() == 1) {
                    for (int j = 0; j < line.size(); j++) {
                        if (j > 0) {
                            s += " ";
                        }
                        s += line[j];
                    }

                    // pad remaining spaces
                    int remaining = maxWidth - s.size();
                    for (int k = 0; k < remaining; k++) {
                        s += " ";
                    }
                }
                else {
                    int spaces = maxWidth - length;
                    int gaps = line.size() - 1;
                    int even = spaces / gaps;
                    int extra = spaces % gaps;

                    for (int j = 0; j < line.size(); j++) {
                        s += line[j];

                        if (j < gaps) {
                            int count = even;
                            if (j < extra) {
                                count += 1;
                            }

                            for (int k = 0; k < count; k++) {
                                s += " ";
                            }
                        }
                    }
                }
                result.push_back(s);
            }

            return result;
    }
};

int main() {
    Solution solution1;
    vector<string> words1 = {"This", "is", "an", "example", "of", "text", "justification."};
    int maxWidth1 = 16;
    vector<string> result1 = solution1.fullJustify(words1, maxWidth1);
    cout << "Testing Text Justification: " << endl;
    cout << "Input: words = [\"This\", \"is\", \"an\", \"example\", \"of\", \"text\", \"justification.\"], maxWidth = 16" << endl;
    cout << "Output: [";
    for (int i = 0; i < result1.size(); i++) {
        cout << "\"" << result1[i] << "\"";
        if (i < result1.size() - 1) {
            cout << ",";
        }
    }
    cout << "]" << endl;

    Solution solution2;
    vector<string> words2 = {"What","must","be","acknowledgment","shall","be"};
    int maxWidth2 = 16;
    vector<string> result2 = solution2.fullJustify(words2, maxWidth2);
    cout << "" << endl;
    cout << "Input: words = [\"What\",\"must\",\"be\",\"acknowledgment\",\"shall\",\"be\"], maxWidth = 16" << endl;
    cout << "Output: [";
    for (int i = 0; i < result2.size(); i++) {
        cout << "\"" << result2[i] << "\"";
        if (i < result2.size() - 1) {
            cout << ",";
        }
    }
    cout << "]" << endl;

    Solution solution3;
    vector<string> words3 = {"Science","is","what","we","understand","well","enough","to","explain","to","a","computer.","Art","is","what","we","do","with","what","we","understand."};
    int maxWidth3 = 20;
    vector<string> result3 = solution3.fullJustify(words3, maxWidth3);
    cout << "" << endl;
    cout << "Input: words = [\"Science\",\"is\",\"what\",\"we\",\"understand\",\"well\",\"enough\",\"to\",\"explain\",\"to\",\"a\",\"computer.\",\"Art\",\"is\",\"what\",\"we\",\"do\",\"with\",\"what\",\"we\",\"understand.\"], maxWidth = 20" << endl;
    cout << "Output: [";
    for (int i = 0; i < result3.size(); i++) {
        cout << "\"" << result3[i] << "\"";
        if (i < result3.size() - 1) {
            cout << ",";
        }
    }
    cout << "]" << endl;

    return 0;
}

File: Hard/text_justification/text_justification.py
from typing import List
import json

def fullJustify(words: List[str], maxWidth: int) -> List[str]:
    result = []
    i = 0

    while i < len(words):
        line = []
        length = 0
        while (i < len(words) and length + len(words[i]) + len(line) <= maxWidth):
            line.append(words[i])
            length += len(words[i])
            i += 1
        if i == len(words) or len(line) == 1: #If last line or only one word in the line, left justify
            s = " ".join(line)
            s += " " * (maxWidth - len(s))
        else:
            spaces = maxWidth - length
            gaps = len(line) - 1
            even_spaces = spaces // gaps
            extra_spaces = spaces % gaps

            s = ""
            for j in range(len(line)):
                s += line[j]
                if j < gaps:
                    if j < extra_spaces:
                        s += " " * (even_spaces + 1)
                    else:
                        s += " " * even_spaces
        result.append(s)
    return result

if __name__ == "__main__":
    print("Testing fullJustify function:")
    print("For the input [\"This\", \"is\", \"an\", \"example\", \"of\", \"text\", \"justification.\"] and maxWidth 16, the output is:", json.dumps(fullJustify(["This", "is", "an", "example", "of", "text", "justification."], 16)))
    print("For the input [\"What\",\"must\",\"be\",\"acknowledgment\",\"shall\",\"be\"] and maxWidth 16, the output is:", json.dumps(fullJustify(["What","must","be","acknowledgment","shall","be"], 16)))
    print("For the input [\"Science\",\"is\",\"what\",\"we\",\"understand\",\"well\",\"enough\",\"to\",\"explain\",\"to\",\"a\",\"computer.\",\"Art\",\"is\",\"everything\",\"else\",\"we\",\"do\"] and maxWidth 20, the output is:", json.dumps(fullJustify(["Science","is","what","we","understand","well","enough","to","explain","to","a","computer.","Art","is","everything","else","we","do"], 20)))